src/EditorFeatures/Test/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs

```diff
@@ -152,6 +153,146 @@ public abstract class AbstractCodeActionOrUserDiagnosticTest
+                : await TestWorkspace.CreateAsync(initialMarkup))
+            {
+                var codeActions = await GetCodeActionsAsync(workspace, fixAllActionEquivalenceKey: null);
+                await TestAddDocument(workspace, expectedMarkup, index, expectedContainers, expectedDocumentName,
+                    codeActions, compareTokens);
+            }
+        }
+
+        private async Task<Tuple<Solution, Solution>> TestAddDocument(
+            TestWorkspace workspace,
+            string expectedMarkup,
+            int index,
+            IList<string> expectedFolders,
+            string expectedDocumentName,
+            IList<CodeAction> actions,
+            bool compareTokens)
+        {
+            var operations = await VerifyInputsAndGetOperationsAsync(index, actions);
+            return await TestAddDocument(
+                workspace,
+                expectedMarkup,
+                operations,
+                hasProjectChange: false,
+                modifiedProjectId: null,
+                expectedFolders: expectedFolders,
+                expectedDocumentName: expectedDocumentName,
+                compareTokens: compareTokens);
+        }
+
+        protected async Task<Tuple<Solution, Solution>> TestAddDocument(
+            TestWorkspace workspace,
+            string expected,
+            IEnumerable<CodeActionOperation> operations,
+            bool hasProjectChange,
+            ProjectId modifiedProjectId,
+            IList<string> expectedFolders,
+            string expectedDocumentName,
+            bool compareTokens)
+        {
+            var appliedChanges = ApplyOperationsAndGetSolution(workspace, operations);
+            var oldSolution = appliedChanges.Item1;
+            var newSolution = appliedChanges.Item2;
+
+            Document addedDocument = null;
+            if (!hasProjectChange)
+            {
+                addedDocument = SolutionUtilities.GetSingleAddedDocument(oldSolution, newSolution);
+            }
+            else
+            {
+                Assert.NotNull(modifiedProjectId);
+                addedDocument = newSolution.GetProject(modifiedProjectId).Documents.SingleOrDefault(doc => doc.Name == expectedDocumentName);
+            }
+
+            Assert.NotNull(addedDocument);
+
+            AssertEx.Equal(expectedFolders, addedDocument.Folders);
+            Assert.Equal(expectedDocumentName, addedDocument.Name);
+            if (compareTokens)
+            {
+                TokenUtilities.AssertTokensEqual(
+                    expected, (await addedDocument.GetTextAsync()).ToString(), GetLanguage());
+            }
+            else
+            {
+                Assert.Equal(expected, (await addedDocument.GetTextAsync()).ToString());
+            }
+
+            var editHandler = workspace.ExportProvider.GetExportedValue<ICodeActionEditHandlerService>();
+            if (!hasProjectChange)
```

Comment:
No, I'm moving this code down from `AbstractUserDiagnosticTest` to `AbstractCodeActionOrUserDiagnosticTest` as thats what this should belong.